File: puppyscript/puppy/structure.py
import re

from .ids import next_name, with_cmd, break_cmd, firsts, with_stmt_line, stmt_add_brc, WORD


class Structure:
    content: str

    def __init__(self, file: str):
        with open(file, "r", encoding="utf-8") as f:
            self.content = f.read()

    def scan_if(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.match(r"(\s*)else\s*:", line)
            if tmp is None:
                tmp = re.match(r"(\s*)((if|elif)(_not)?)\s+(.+):", line)
                if tmp is None:
                    result.append(line)
                else:
                    result.append(with_stmt_line(tmp.group(1), tmp.group(5)))
                    result.append(tmp.group(1) + tmp.group(2) + ":")
            else:
                result.append("%s %s" % (tmp.group(1) + with_cmd, "true"))
                result.append(tmp.group(1) + "elif:")
        self.content = "\n".join(result)

    def scan_for(self):
        result = list()
        for_name: tuple[str, str, str, str] | None = None
        outer_indent = inner_indent = None
        succeeded = False
        for line in self.content.split("\n"):
            if for_name is None:
                tmp = re.fullmatch(rf"(\s*)for\s+(.+?)\s*,\s*(.+?)\s*,\s*(.+?)\s*:\s*", line)
                if tmp is not None:
                    outer_indent = tmp.group(1)
                    for_name = (tmp.group(2), tmp.group(3), tmp.group(4), None)
                tmp = re.fullmatch(rf"(\s*)for\s+({WORD})\s+in\s+(.+?):\s*", line)
                if tmp is not None:
                    outer_indent = tmp.group(1)
                    tmp_name = next_name("FOR")
                    result.append(outer_indent + tmp_name + " = " + tmp.group(3))
                    for_name = (tmp.group(2) + " = call " + tmp_name + ".beg",
                                "call " + tmp_name + ".cond",
                                tmp.group(2) + " = call " + tmp_name + ".next",
                                tmp_name)
                if for_name is not None:
                    succeeded = True
                    result.append(outer_indent + for_name[0])
                    result.append(outer_indent + "while " + for_name[1] + ":")
                    continue
                result.append(line)
            else:
                cur_indent = firsts(line)
                if inner_indent is None:
                    inner_indent = cur_indent
                else:
                    if len(cur_indent) < len(inner_indent):
                        result.append(inner_indent + for_name[2])
                        if for_name[3] is not None:
                            result.append(outer_indent + "~" + for_name[3])
                        for_name = None
                        outer_indent = inner_indent = None
                result.append(line)
        self.content = "\n".join(result)
        return succeeded

    def scan_all_for(self):
        while self.scan_for():
            ...

    def scan_while(self):
        result = list()
        while_name = None
        stack = list()
        outer_indent = inner_indent = None
        for line in self.content.split("\n"):
            if while_name is False:
                result.append(line)
            elif while_name is not None:
                stack.append(line)
                if inner_indent is None:
                    inner_indent = firsts(line)
                elif len(firsts(line)) < len(inner_indent):
                    stack.pop()
                    stack[0] = outer_indent + "if:" + stack[0].strip()
                    stack.insert(0, with_stmt_line(outer_indent, while_name))
                    stack.append("%s %s" % (inner_indent +
                                            break_cmd, stmt_add_brc(while_name)))

                    result.extend(stack)
                    result.append(line)
                    while_name = False
            else:
                tmp = re.match(r"(\s*)while(_not)?\s+(.+):", line)
                if tmp is None:
                    result.append(line)
                else:
                    while_name = tmp.group(3)
                    outer_indent = tmp.group(1)
                    if tmp.group(2) is None:
                        stack = [outer_indent + "loop:"]
                    else:
                        stack = [outer_indent + "loop_not:"]

        self.content = "\n".join(result)
        return while_name is not None

    def scan_all_while(self):
        while self.scan_while():
            ...

    def scan_match_case(self):
        result = list()
        for line in self.content.split("\n"):
            tmp = re.fullmatch(r"(\s*)(match|case)\s+(.+?)\s*:\s*", line)
            if tmp is None:
                result.append(line)
            else:
                result.append(with_stmt_line(tmp.group(1), tmp.group(3)))
                result.append(tmp.group(1) + tmp.group(2) + ":")
        self.content = "\n".join(result)

    def work(self, output_name: str) -> str:
        self.scan_if()
        self.scan_all_for()
        self.scan_all_while()
        self.scan_match_case()
        output = output_name + ".struct"
        with open(output, "w", encoding="utf-8") as f:
            f.write(self.content)
        return output

File: puppyscript/cpp-puppy/structure.cpp
//
// Created by impodog on 7/30/2023.
//
#include "processor.h"

std::regex RE_ELSE(R"((\s*)else\s*:)"),
        RE_IF(R"((\s*)((if|elif)(_not)?)\s+(.+):)"),
        RE_FOR_BASIC(R"((\s*)for\s+(.+?)\s*,\s*(.+?)\s*,\s*(.+?)\s*:\s*)"),
        RE_FOR_RANGE("(\\s*)for\\s+(" WORD ")\\s+in\\s+(.+?):\\s*"),
        RE_WHILE(R"((\s*)while\s+(.+):)"),
        RE_MATCH_CASE(R"((\s*)(match|case)\s+(.+):\s*)");

class Structure {
public:
    std::string content;

    explicit Structure(const path &file) {
        content = read(file);
    }

    void scan_if() {
        string_list result;
        auto lines = split(content, '\n');
        for (const auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_ELSE)) {
                result.push_back(results[1].str() + WITH_STMT " true");
                result.push_back(results[1].str() + "elif:");
            } else if (std::regex_match(line, results, RE_IF)) {
                result.push_back(add_with(results[1].str(), results[5].str()));
                result.push_back(results[1].str() + results[2].str() + ":");
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    bool scan_for() {
        string_list result;
        std::string *for_name = nullptr;
        std::string outer_indent, inner_indent;
        bool changed = false;

        auto lines = split(content, '\n');
        for (const auto &line: lines) {
            if (for_name) {
                auto cur_indent = indent_of(line);
                if (inner_indent.empty()) {
                    inner_indent = cur_indent;
                } else {
                    if (cur_indent.size() < inner_indent.size()) {
                        result.push_back(inner_indent + for_name[2]);
                        if (!for_name[3].empty()) {
                            result.push_back(outer_indent + "~" + for_name[3]);
                        }
                        delete[] for_name;
                        for_name = nullptr;
                        outer_indent.clear();
                        inner_indent.clear();
                    }
                }
                result.push_back(line);
            } else {
                std::smatch results;
                if (std::regex_match(line, results, RE_FOR_BASIC)) {
                    outer_indent = results[1].str();
                    for_name = new std::string[4];
                    for_name[0] = results[2].str();
                    for_name[1] = results[3].str();
                    for_name[2] = results[4].str();
                    for_name[3] = "";
                } else if (std::regex_match(line, results, RE_FOR_RANGE)) {
                    outer_indent = results[1].str();
                    auto tmp_name = globals->next_name("FOR");
                    result.push_back(outer_indent + tmp_name + " = " + results[3].str());
                    for_name = new std::string[4];
                    for_name[0] = results[2].str() + " = call " + tmp_name + ".beg";
                    for_name[1] = "call " + tmp_name + ".cond";
                    for_name[2] = results[2].str() + " = call " + tmp_name + ".next";
                    for_name[3] = tmp_name;
                }
                if (for_name) {
                    changed = true;
                    result.push_back(outer_indent + for_name[0]);
                    result.push_back(outer_indent + "while " + for_name[1] + ":");
                    continue;
                }
                result.push_back(line);
            }
        }
        content = join(result, '\n');
        return changed;
    }

    void scan_all_for() {
        while (scan_for());
    }

    bool scan_while() {
        string_list result;
        std::string while_name;
        bool end_scanning = false;
        std::list<std::string> stack;
        std::string outer_indent, inner_indent;

        auto lines = split(content, '\n');
        for (const auto &line: lines) {
            if (while_name.empty()) {
                std::smatch results;
                if (std::regex_match(line, results, RE_WHILE)) {
                    while_name = results[2].str();
                    outer_indent = results[1].str();
                    stack = {outer_indent + "loop:"};
                } else {
                    result.push_back(line);
                }
            } else if (end_scanning) {
                result.push_back(line);
            } else {
                stack.push_back(line);
                if (inner_indent.empty()) {
                    inner_indent = indent_of(line);
                } else if (indent_of(line).size() < inner_indent.size()) {
                    stack.pop_back();
                    stack.front() = outer_indent + "if:" + strip(stack.front());
                    stack.insert(stack.begin(), add_with(outer_indent, while_name));
                    stack.push_back(inner_indent + "break " + add_brackets(while_name));
                    result.insert(result.end(), stack.begin(), stack.end());
                    result.push_back(line);
                    end_scanning = true;
                }
            }
        }
        content = join(result, '\n');
        return !while_name.empty();
    }

    void scan_all_while() {
        while (scan_while());
    }

    void scan_match_case() {
        string_list result;
        auto lines = split(content, '\n');
        for (const auto &line: lines) {
            std::smatch results;
            if (std::regex_match(line, results, RE_MATCH_CASE)) {
                result.push_back(add_with(results[1].str(), results[3].str()));
                result.push_back(results[1].str() + results[2].str() + ":");
            } else {
                result.push_back(line);
            }
        }
        content = join(result, '\n');
    }

    std::string work(const std::string &output_name) {
        scan_if();
        scan_all_for();
        scan_all_while();
        scan_match_case();

        std::string output = output_name + ".struct.puppy";
        write(output, content);
        return output;
    }
};

PROCESSOR(structure) {
    Structure structure(file);
    return structure.work(output_name);
}